File: packages/api/src/menu.ts
import { invoke, listen } from "./transport.js";

/** A menu entry. `type` defaults to `normal`. */
export type MenuItem =
  | {
      type?: "normal";
      id: string;
      label: string;
      enabled?: boolean;
      /** e.g. `CmdOrCtrl+S`, `Alt+Shift+F4`. */
      accelerator?: string;
    }
  | {
      type: "checkbox";
      id: string;
      label: string;
      checked?: boolean;
      enabled?: boolean;
      accelerator?: string;
    }
  | { type: "submenu"; label: string; items: MenuItem[]; enabled?: boolean }
  | { type: "separator" }
  | { type: "predefined"; item: PredefinedMenuItem; label?: string };

export type PredefinedMenuItem =
  | "separator"
  | "copy"
  | "cut"
  | "paste"
  | "selectAll"
  | "undo"
  | "redo"
  | "minimize"
  | "maximize"
  | "fullscreen"
  | "hide"
  | "hideOthers"
  | "showAll"
  | "closeWindow"
  | "quit"
  | "about"
  | "services"
  | "bringAllToFront";

/**
 * The application menu.
 *
 * On macOS this is the menu bar, and it is not optional decoration: the
 * predefined `copy`, `paste`, `undo` and `selectAll` items are what make
 * those keyboard shortcuts work at all. An app that sets no menu has no
 * working Cmd-C.
 *
 * ```ts
 * await menu.set([
 *   { type: "submenu", label: "File", items: [
 *     { id: "new", label: "New", accelerator: "CmdOrCtrl+N" },
 *     { type: "separator" },
 *     { type: "predefined", item: "quit" },
 *   ]},
 *   { type: "submenu", label: "Edit", items: [
 *     { type: "predefined", item: "undo" },
 *     { type: "predefined", item: "copy" },
 *     { type: "predefined", item: "paste" },
 *   ]},
 * ]);
 *
 * menu.onClick(({ id }) => { if (id === "new") newDocument(); });
 * ```
 */
export const menu = {
  set: (items: MenuItem[]) =>
    invoke<null>("menu.set", { items: normalize(items) }),
  remove: () => invoke<null>("menu.remove"),

  setEnabled: (id: string, enabled: boolean) =>
    invoke<null>("menu.setEnabled", { id, enabled }),
  setLabel: (id: string, label: string) =>
    invoke<null>("menu.setLabel", { id, label }),
  setChecked: (id: string, checked: boolean) =>
    invoke<null>("menu.setChecked", { id, checked }),
  isChecked: (id: string) => invoke<boolean>("menu.isChecked", { id }),

  /**
   * Show a context menu. Position is in logical pixels relative to the
   * window; the cursor is used when it is omitted.
   */
  popup: (
    items: MenuItem[],
    options: { x?: number; y?: number; label?: string } = {},
  ) => invoke<null>("menu.popup", { ...options, items: normalize(items) }),

  /** Fires for any item with an `id`, in the app menu, a tray menu or a popup. */
  onClick: (handler: (event: { id: string }) => void) =>
    listen<{ id: string }>("menu.click", handler),
};

/**
 * Fill in the `type` the runtime insists on.
 *
 * `{ id, label }` is the item people write nine times out of ten, and the
 * runtime's parser is a tagged union with no default - which is exactly what
 * makes a misspelled `type` an error instead of a silent normal item.
 */
export function normalize(items: MenuItem[]): MenuItem[] {
  return items.map((item) => {
    const entry = { type: "normal", ...item } as MenuItem & {
      items?: MenuItem[];
    };
    return entry.items ? { ...entry, items: normalize(entry.items) } : entry;
  });
}

File: runtime/src/api/menu.rs
//! `menu.*` - the application menu, and context menus.
//!
//! On macOS this is the menu bar, and it is not decoration: without the
//! predefined `copy`, `paste`, `undo` and `selectAll` items in it, those
//! keyboard shortcuts do not work anywhere in the application. A Vantail app
//! that wants working Cmd-C sets a menu.

use serde::Deserialize;
use serde_json::{json, Value};

use crate::chrome::menu::MenuSpec;
use crate::error::{ApiError, ApiResult};
use crate::ipc::Request;
use crate::state::MainCtx;

#[derive(Deserialize)]
struct Items {
    items: Vec<MenuSpec>,
}

#[derive(Deserialize)]
struct SetEnabled {
    id: String,
    enabled: bool,
}

#[derive(Deserialize)]
struct SetChecked {
    id: String,
    checked: bool,
}

#[derive(Deserialize)]
struct SetLabel {
    id: String,
    label: String,
}

#[derive(Deserialize)]
struct Id {
    id: String,
}

#[derive(Deserialize, Default)]
#[serde(rename_all = "camelCase", default)]
struct Popup {
    items: Vec<MenuSpec>,
    /// Where to show it, in logical pixels relative to the window. The cursor
    /// position when omitted.
    x: Option<f64>,
    y: Option<f64>,
    label: Option<String>,
}

pub fn dispatch(ctx: &mut MainCtx<'_>, method: &str, params: Value) -> ApiResult {
    ctx.rt
        .permissions
        .require(ctx.rt.permissions.menu, method)?;

    match method {
        "menu.set" => {
            let Items { items } = Request::params(method, params)?;
            ctx.chrome.set_app_menu(&items)?;
            // Windows attaches the menu per window rather than per app.
            for label in ctx.windows.labels() {
                if let Some(entry) = ctx.windows.get(&label) {
                    ctx.chrome.attach(&entry.window);
                }
            }
            Ok(Value::Null)
        }

        "menu.remove" => {
            ctx.chrome.remove_app_menu();
            Ok(Value::Null)
        }

        "menu.setEnabled" => {
            let SetEnabled { id, enabled } = Request::params(method, params)?;
            item(ctx, &id)?.set_enabled(enabled);
            Ok(Value::Null)
        }

        "menu.setLabel" => {
            let SetLabel { id, label } = Request::params(method, params)?;
            item(ctx, &id)?.set_label(&label);
            Ok(Value::Null)
        }

        "menu.setChecked" => {
            let SetChecked { id, checked } = Request::params(method, params)?;
            if !item(ctx, &id)?.set_checked(checked) {
                return Err(ApiError::invalid_params(format!(
                    "`{id}` is not a checkbox item"
                )));
            }
            Ok(Value::Null)
        }

        "menu.isChecked" => {
            let Id { id } = Request::params(method, params)?;
            item(ctx, &id)?
                .is_checked()
                .map(|checked| json!(checked))
                .ok_or_else(|| ApiError::invalid_params(format!("`{id}` is not a checkbox item")))
        }

        "menu.popup" => popup(ctx, params),

        _ => Err(ApiError::unknown_method(method)),
    }
}

fn item<'a>(
    ctx: &'a MainCtx<'_>,
    id: &str,
) -> Result<&'a crate::chrome::menu::ItemHandle, ApiError> {
    ctx.chrome.item(id).ok_or_else(|| {
        ApiError::new(
            crate::error::code::NOT_FOUND,
            format!("No menu item with id `{id}`"),
        )
    })
}

#[allow(unused_variables)]
fn popup(ctx: &mut MainCtx<'_>, params: Value) -> ApiResult {
    let params: Popup = Request::params("menu.popup", params)?;
    if params.items.is_empty() {
        return Err(ApiError::invalid_params("`items` cannot be empty"));
    }

    let label = params.label.unwrap_or_else(|| ctx.source.to_string());
    let entry = ctx.windows.require(&label)?;

    // A throwaway menu: the platform call below is modal, so it stays alive
    // for exactly as long as it is on screen.
    let mut handles = std::collections::HashMap::new();
    let menu = crate::chrome::menu::build(&params.items, &mut handles)
        .map_err(ApiError::invalid_params)?;

    // Annotated rather than inferred: only the platform-specific calls below
    // pin this type, so on a platform where none of them compile there is
    // nothing left to infer it from.
    let position: Option<muda::dpi::Position> = match (params.x, params.y) {
        (Some(x), Some(y)) => Some(muda::dpi::LogicalPosition::new(x, y).into()),
        _ => None,
    };

    #[cfg(target_os = "macos")]
    {
        use muda::ContextMenu;
        use tao::platform::macos::WindowExtMacOS;
        // Safety: the view comes straight from the window we just looked up,
        // and the call returns before this function does.
        unsafe {
            menu.show_context_menu_for_nsview(entry.window.ns_view(), position);
        }
        return Ok(Value::Null);
    }

    #[cfg(target_os = "windows")]
    {
        use muda::ContextMenu;
        use tao::platform::windows::WindowExtWindows;
        // Safety: the handle belongs to the window we just looked up.
        unsafe {
            menu.show_context_menu_for_hwnd(entry.window.hwnd(), position);
        }
        return Ok(Value::Null);
    }

    #[allow(unreachable_code)]
    Err(ApiError::unsupported(
        "Context menus are not implemented on this platform yet",
    ))
}
